include/swift/SIL/SILLocation.h

```diff
@@ -15,29 +15,67 @@
 
 #include "llvm/ADT/PointerUnion.h"
 #include "swift/Basic/SourceLoc.h"
-#include "swift/Basic/SourceManager.h"
+#include "swift/SIL/SILAllocated.h"
 #include "swift/AST/TypeAlignments.h"
 
 #include <cstddef>
 #include <type_traits>
 
 namespace swift {
 
-class SourceLoc;
 class ReturnStmt;
 class BraceStmt;
 class AbstractClosureExpr;
 class AbstractFunctionDecl;
+class SILModule;
+class SourceManager;
 
-/// This is a pointer to the AST node that a SIL instruction was
-/// derived from. This may be null if AST information is unavailable or
-/// stripped.
-///
-/// FIXME: This should eventually include inlining history, generics
-/// instantiation info, etc (when we get to it).
+/// Represents a source location.
 ///
+/// In most cases this is a pointer to an AST node from which the source
+/// location can be retrieved from. In case the location refers to a parsed
+/// or deserialized SIL file, the source location is represented differently.
+/// For details see SILLocation::Storage.
 class SILLocation {
-private:
+public:
+  enum LocationKind : uint8_t {
+    RegularKind = 1,
+    ReturnKind = 2,
+    ImplicitReturnKind = 3,
+    InlinedKind = 4,
+    MandatoryInlinedKind = 5,
+    CleanupKind = 6,
+    ArtificialUnreachableKind = 7
+  };
+
+  /// Describes a position in a source file by explicitly storing the file name,
+  /// line and column.
+  ///
+  /// This is used for parsed locations from a SIL file, for "-gsil" (see
+  /// SILDebugInfoGenerator) and for the "compiler-generated" singleton location.
+  /// In future we might also use it for de-serialized locations from a
+  /// swiftmodule file.
+  struct FilePosition : public SILAllocated<FilePosition> {
+    unsigned Line;
+    unsigned Column;
```

Comment:
LLVM's DILocation restricts Columns to 16 bits. This probably doesn't save real space due to alignment requirements, but it would document the restrictions better perhaps?